src/dbnode/storage/namespace/schema_registry.go

```diff
@@ -0,0 +1,177 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package namespace
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/m3db/m3/src/x/ident"
+	xclose "github.com/m3db/m3/src/x/close"
+	xwatch "github.com/m3db/m3/src/x/watch"
+	"go.uber.org/zap"
+)
+
+type schemaRegistry struct {
+	sync.RWMutex
+
+	protoEnabled bool
+	logger *zap.Logger
+	registry map[string]xwatch.Watchable
+}
+
+func NewSchemaRegistry(protoEnabled bool, logger *zap.Logger) SchemaRegistry {
+	return newSchemaRegistry(protoEnabled, logger)
+}
+
+func newSchemaRegistry(protoEnabled bool, logger *zap.Logger) SchemaRegistry {
+	return &schemaRegistry{
+		protoEnabled: protoEnabled,
+		logger: logger,
+		registry: make(map[string]xwatch.Watchable),
+	}
+}
+
+func (sr *schemaRegistry) SetSchemaHistory(id ident.ID, history SchemaHistory) error {
+	if !sr.protoEnabled {
+		if sr.logger != nil {
+			sr.logger.Warn("proto is not enabled, can not update schema registry",
+				zap.Stringer("namespace", id))
+		}
+		return nil
+	}
+
+	if _, ok := history.GetLatest(); !ok {
+		return fmt.Errorf("can not set empty schema history for %v", id.String())
+	}
+
+	sr.Lock()
+	defer sr.Unlock()
+
+	// TODO [haijun] use generated map for optimized map lookup.
+	current, ok := sr.registry[id.String()]
```

Comment:
nit: maybe pull out `id.String()` into a var so it only gets called once?

i.e. `key := id.String()` and then use `key` in map lookups and sets below?